exp/query/lsm_fetcher.go

```diff
@@ -0,0 +1,177 @@
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package query
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"github.com/weaviate/weaviate/adapters/repos/db/lsmkv"
+	"github.com/weaviate/weaviate/entities/cyclemanager"
+)
+
+const (
+	defaultLSMRoot = "lsm"
+
+	// NOTE(kavi): using fixed nodeName that offloaded tenant under that prefix on object storage.
+	// TODO(kavi): Make it dynamic.
+	nodeName = "weaviate-0"
+)
+
+// LSMFetcher fetches the tenant's LSM data using given `upstream` without or without local cache.
+type LSMFetcher struct {
+	cache    LSMCache
+	upstream LSMDownloader
+	log      logrus.FieldLogger
+
+	basePath string
+}
+
+// LSMDownloader is usually some component that used to download tenant's data.
+type LSMDownloader interface {
+	DownloadToPath(ctx context.Context, collection, tenant, nodeName, path string) error
+}
+
+// LSMCache provides caching of tenant's LSM data.
+type LSMCache interface {
+	Tenant(collection, tenantID string) (*TenantCache, error)
+	AddTenant(collection, tenantID string, version uint64) error
+	BasePath() string
+}
+
+// NewLSMFetcher creates LSMFetcher without cache.
+// Every call is forwarded to given `upstream` to fetch the LSM data for the tenant.
+func NewLSMFetcher(basePath string, upstream LSMDownloader, log logrus.FieldLogger) *LSMFetcher {
+	return NewLSMFetcherWithCache(basePath, upstream, nil, log)
+}
+
+// NewLSMFetcherWithCache creates LSMFetcher with cache.
+func NewLSMFetcherWithCache(basePath string, upstream LSMDownloader, cache LSMCache, log logrus.FieldLogger) *LSMFetcher {
+	return &LSMFetcher{
+		basePath: basePath,
+		cache:    cache,
+		log:      log,
+		upstream: upstream,
+	}
+}
+
+// Fetch try to fetch LSM data for the given (collection,tenant).
+// It works in following modes
+// 1. If cache is nil, fetch from upstream.
+// 2. If cache is enabled and not found in cache, fetch from upstream.
+// 3. If cache is enabled and found in the cache but local version is outdated, fetch from the upstream and remove old version.
+// 4. if cache is enabled, found in the cache and local version is up to date, return from the cache.
+func (l *LSMFetcher) Fetch(ctx context.Context, collection, tenant string, version uint64) (*lsmkv.Store, string, error) {
+	var (
+		basePath          = l.basePath
+		tenantPath string = ""
+		download   bool   = true
+		oldVersion uint64
+	)
+
+	if l.cache != nil {
+		basePath = l.cache.BasePath()
+		c, err := l.cache.Tenant(collection, tenant)
+		if err == nil && version != 0 && c.Version >= version {
+			tenantPath = c.AbsolutePath()
+			download = false
+		}
+		if err == nil && version != 0 && c.Version < version {
+			// oldVersion is used to cleanup previous version of tenant data.
+			oldVersion = c.Version
+		}
+		if err != nil && !errors.Is(err, ErrTenantNotFound) {
```

Comment:
NIT: we check both above that err == nil, we can just flip the condition to first check `if err != nil` and then have the other conditions. Makes the cody slightly easier to read and more go-idiomatic by having the err check first.

Reply:
Done